### PR TITLE
fix ApiGateway ASF request matching

### PR DESCRIPTION
**Pull request by alexrashed**
This PR addresses an issue which arose in https://github.com/localstack/localstack/pull/5731, where dynamically created API Gateway resources were (falsely) matching a service operation:
Requests against `/restapis/cmqinv79uh/local/_user_request_/` were wrongly matched as requests for the the API Gateway operation `GetRestApi`.

It fixes the issue by heavily refactoring / extracting the detection of the ASF service operations for REST protocols.
- The new `RestServiceOperationRouter` heavily uses the request matching logic of [Werkzeug](https://github.com/pallets/werkzeug).
- It clearly separates the path, method, query, and header matching and now strictly enforces path boundaries (i.e. requests to subpaths are not falsely matched anymore).
- It changed the handling of greedy parameters which builds on the assumption that:
  - A greedy path parameter is explicitly marked as such in the specification by using the `+` notation ([as described here in the Smithy HTTP binding specs](https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#greedy-labels)). These parameters can contain slashes in the value.
    - f.e. `/foo/bar/baz.txt` matches `/{Bucket}/{Key+}` where `Bucket=foo` and `Key=bar/baz.txt`.
  - A path parameter which is _not_ marked as greedy and contains slashes is properly url encoded.
    - f.e. `/tags/arn%3Aaws%3Aqldb%3Aeu-central-1%3A000000000000%3Aledger%2Fc-c67c827a` matches `/tags/{resourceArn}`.
    - This is how `botocore` also encodes the path parameter when sending the requests: [botocore/serialize.py#L469-L484](https://github.com/boto/botocore/blob/3c4bd9bb7c12869be49847adeb949babd1fa5a86/botocore/serialize.py#L469-L484)

Props go to @thrau for the initial implementation of this new operation router! 🥇

### CODE REVIEW

File: tests/unit/aws/protocol/test_serializer.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 from botocore.parsers import ResponseParser, create_parser
 from dateutil.tz import tzlocal, tzutc
 
-from localstack.aws.api import CommonServiceException, ServiceException
+from localstack.aws.api import CommonServiceException, HttpResponse, ServiceException
```
Comment (thrau) [inline tests/unit/aws/protocol/test_serializer.py:10]: nit: can we import `localstack.http.Response` instead?

File: localstack/aws/protocol/router.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        :return: matching fine-grained rule
+        :raises: NotFound if none of the fine-grained rules matches
+        """
+        try:
+            # return the first matching rule or raise a NotFound error if no rule matches
+            return next(
+                filter(lambda rule: rule.matches(request.args, request.headers), self.rules)
+            )
+        except StopIteration:
+            raise NotFound()
```
Comment (thrau) [inline localstack/aws/protocol/router.py:140]: FWIW, i've seen this pattern several times now and I have to say I'm not super fond of it. Isn't this version a lot easier to read?:
```python
for rule in self.rules:
    if rule.matches(request.args, request.headers):
        return rule
raise NotFound()
```

plus it avoids the extra `StopIteration` except overhead.
is there any benefit to the `next`,`filter`,`lambda` nesting i may be missing?

File: localstack/aws/protocol/router.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+class _HttpOperation(NamedTuple):
+    operation: OperationModel
+    path: str
+    method: str
+    query_args: Mapping[str, List[str]]
+    header_args: List[str]
+
+    @staticmethod
+    def from_operation(op: OperationModel) -> "_HttpOperation":
+        """Useful intermediary representation of the 'http' block of an operation to make code cleaner"""
```
Comment (thrau) [inline localstack/aws/protocol/router.py:74]: nit: maybe move this comment up to the class doc, and add some concrete method doc here on how it works

File: localstack/aws/protocol/router.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        required_query_args: Optional[Mapping[str, List[Any]]],
+        required_header_args: Optional[List[str]],
+    ) -> None:
+        super().__init__()
+        self.endpoint = endpoint
+        self.required_query_args = required_query_args or {}
+        self.required_header_args = required_header_args or []
+        self.match_score = 1 + len(self.required_query_args) + len(self.required_header_args)
+
+    def matches(self, query_args: "MultiDict[str, str]", headers: Headers) -> bool:
```
Comment (thrau) [inline localstack/aws/protocol/router.py:38]: any reason we are using the double quotes around the `MultiDict` type hint here?

File: localstack/aws/protocol/router.py (hunk 2 above)
Comment (alexrashed) [reply localstack/aws/protocol/router.py:140]: I guess it's question of taste. I like the functional approach a bit more, but I see your point (and mine is highly subjective). I'll change it.

**Pull request merged**